src/Mvc/Mvc.Core/src/Formatters/InputFormatter.cs

```diff
@@ -98,8 +99,11 @@ public virtual Task<InputFormatterResult> ReadAsync(InputFormatterContext contex
             throw new ArgumentNullException(nameof(context));
         }
 
-        var request = context.HttpContext.Request;
-        if (request.ContentLength == 0)
+        var hasBody = context.HttpContext.Features.Get<IHttpRequestBodyDetectionFeature>()?.CanHaveBody;
+        // In case the feature is not registered
+        hasBody ??= context.HttpContext.Request.ContentLength is not null && context.HttpContext.Request.ContentLength != 0;
+
+        if (hasBody == false)
```

Comment:
```suggestion
        var canHaveBody = context.HttpContext.Features.Get<IHttpRequestBodyDetectionFeature>()?.CanHaveBody
        // In case the feature is not registered
            ??  context.HttpContext.Request.ContentLength != 0;

        if (!canHaveBody)
```

We don't want the `context.HttpContext.Request.ContentLength is not null` check because we shouldn't assume the lack of a `ContentLength` means there is no body.

Reply:
I read the `CanHaveBody` property and assumed incorrectly that and I agree with you here, even more after reading the spec https://httpwg.org/specs/rfc7230.html#header.content-length. 

We are only manually checking when the feature is not registered, that is unusual but possible, do you think is OK just be based on the `content-length <> 0` or should we check the `Transfer-Encoding` as explained in the spec?

Reply:
The lack of `Transfer-Encoding` or `Content-Length` header wouldn't be enough either for HTTP/2 and HTTP/3 requests. I think just relying on the `IHttpRequestBodyDetectionFeature` is enough. If the feature is missing and there's a `Content-Length: 0` header, I'm fine saying there must not be a request body, but I wouldn't make it more complicated than that.